File: src/installers/cursor.ts
import { existsSync } from "node:fs";
import { CURSOR_HOOKS_PATH } from "../config.ts";
import {
  type InstallResult,
  isPosthookCommand,
  posthookCommandFor,
  readJsonOrEmpty,
  writeJsonAtomic,
} from "./base.ts";

const AGENT_SLUG = "cursor";
const HOOK_TYPES = ["preToolUse", "postToolUse", "beforeSubmitPrompt", "afterFileEdit"] as const;

interface HookEntry {
  command?: string;
}

export async function detectCursor(): Promise<boolean> {
  const cursorDir = CURSOR_HOOKS_PATH.replace(/\/hooks\.json$/, "");
  return existsSync(cursorDir);
}

export async function installCursorHooks(binaryPath: string): Promise<InstallResult> {
  const path = CURSOR_HOOKS_PATH;
  const before = await readJsonOrEmpty(path);
  const after = JSON.parse(JSON.stringify(before)) as Record<string, unknown>;
  const desiredCmd = posthookCommandFor(binaryPath, AGENT_SLUG);

  if (after.version === undefined) after.version = 1;

  const hooksObj = (after.hooks && typeof after.hooks === "object" && !Array.isArray(after.hooks)
    ? (after.hooks as Record<string, unknown>)
    : {}) as Record<string, HookEntry[]>;
  after.hooks = hooksObj;

  for (const hookType of HOOK_TYPES) {
    const arr: HookEntry[] = Array.isArray(hooksObj[hookType])
      ? (hooksObj[hookType] as HookEntry[])
      : [];

    const idx = arr.findIndex(
      (h) => typeof h.command === "string" && isPosthookCommand(h.command, AGENT_SLUG),
    );
    if (idx === -1) {
      arr.push({ command: desiredCmd });
    } else if (arr[idx]!.command !== desiredCmd) {
      arr[idx] = { command: desiredCmd };
    }
    // Deduplicate.
    const seen = new Set<number>();
    const deduped = arr.filter((h, i) => {
      if (typeof h.command === "string" && isPosthookCommand(h.command, AGENT_SLUG)) {
        if (seen.size > 0) return false;
        seen.add(i);
      }
      return true;
    });

    hooksObj[hookType] = deduped;
  }

  const changed = JSON.stringify(before) !== JSON.stringify(after);
  if (changed) await writeJsonAtomic(path, after);
  return {
    changed,
    path,
    message: changed ? `Cursor: hooks installed in ${path}` : `Cursor: already up to date`,
  };
}

File: internal/installers/cursor.go
package installers

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"github.com/bilanc/posthook/internal/paths"
)

const cursorAgentSlug = "cursor"

var cursorHookTypes = []string{
	"preToolUse", "postToolUse", "beforeSubmitPrompt", "afterFileEdit",
}

func DetectCursor() bool {
	dir := filepath.Dir(paths.CursorHooksPath())
	_, err := os.Stat(dir)
	return err == nil
}

// InstallCursorHooks merges posthook entries into ~/.cursor/hooks.json.
// Cursor's schema is flatter than Claude's: events map directly to an array
// of {command} entries with no matcher layer.
func InstallCursorHooks(binaryPath string) (Result, error) {
	path := paths.CursorHooksPath()
	before, err := ReadJSONOrEmpty(path)
	if err != nil {
		return Result{}, err
	}
	beforeJSON, _ := json.Marshal(before)
	after := cloneJSON(before)
	desiredCmd := PosthookCommandFor(binaryPath, cursorAgentSlug)

	if _, ok := after["version"]; !ok {
		after["version"] = 1
	}

	hooksObj := getOrMakeMap(after, "hooks")
	for _, hookType := range cursorHookTypes {
		arr, _ := hooksObj[hookType].([]any)
		if arr == nil {
			arr = []any{}
		}

		idx := -1
		for i, h := range arr {
			hm, ok := h.(map[string]any)
			if !ok {
				continue
			}
			cmd, _ := hm["command"].(string)
			if IsPosthookCommand(cmd, cursorAgentSlug) {
				idx = i
				break
			}
		}
		if idx == -1 {
			arr = append(arr, map[string]any{"command": desiredCmd})
		} else {
			existing := arr[idx].(map[string]any)
			if existing["command"] != desiredCmd {
				arr[idx] = map[string]any{"command": desiredCmd}
			}
		}

		// Deduplicate any additional copies of our command.
		seen := false
		deduped := make([]any, 0, len(arr))
		for _, h := range arr {
			hm, ok := h.(map[string]any)
			if ok {
				cmd, _ := hm["command"].(string)
				if IsPosthookCommand(cmd, cursorAgentSlug) {
					if seen {
						continue
					}
					seen = true
				}
			}
			deduped = append(deduped, h)
		}
		hooksObj[hookType] = deduped
	}

	afterJSON, _ := json.Marshal(after)
	changed := string(beforeJSON) != string(afterJSON)
	if changed {
		if err := WriteJSONAtomic(path, after); err != nil {
			return Result{}, err
		}
	}
	msg := "Cursor: already up to date"
	if changed {
		msg = fmt.Sprintf("Cursor: hooks installed in %s", path)
	}
	return Result{Changed: changed, Path: path, Message: msg}, nil
}
